### PR TITLE
[KB-186] Xgboost Definition

### PR DESCRIPTION
**Pull request by Ivikhostrup**
Reading guide:
* I've tried to minimize the mathematics a lot since its somewhat convoluted. Is the equation I included redundant or okay?
* What do we think of the level of detail?

### CODE REVIEW

File: report_thesis/src/sections/background.tex [hunk 1]
```diff
@@ -286,6 +286,27 @@ \subsubsection{Natural Gradient Boosting (NGBoost)}
 This iterative process of computing predictions, calculating gradients, fitting weak learners, and updating parameters continues for a predetermined number of iterations or until the model's performance converges.
 
 \subsubsection{XGBoost}
+We now describe XGBoost based on \citet{ChenGuestrin2016}.
```
Comment (Pattrigue) [inline report_thesis/src/sections/background.tex:289]: Cite is broken.

File: report_thesis/src/sections/background.tex [hunk 2]
```diff
@@ -286,6 +286,27 @@ \subsubsection{Natural Gradient Boosting (NGBoost)}
...
+The objective function to be minimized is composed of a convex loss function $L$ that measures the difference between the prediction $\hat{y}_i$ and the target $y_i$, and a regularization term $\Omega$ that penalizes the complexity of the model. 
+A convex loss function has the property of having a single global minimum, which simplifies the optimization process and ensures reliable convergence:
+
+$$
+Obj(\theta) = \sum_{i=1}^{n} L(y_i, \hat{y}_i) + \sum_{k=1}^{K} \Omega(f_k),
+$$
+
+where $f_k$ represents the individual regression trees.
+
+The model is trained in an additive manner, where at each iteration, the objective function is approximated to efficiently compute the gain from adding a new tree.
```
Comment (Pattrigue) [inline report_thesis/src/sections/background.tex:304]:
```suggestion
The model is trained in an additive manner where, at each iteration, the objective function is approximated to efficiently compute the gain from adding a new tree. 
```

File: report_thesis/src/sections/background.tex [hunk 3]
```diff
@@ -286,6 +286,27 @@ \subsubsection{Natural Gradient Boosting (NGBoost)}
...
+
+$$
+Obj(\theta) = \sum_{i=1}^{n} L(y_i, \hat{y}_i) + \sum_{k=1}^{K} \Omega(f_k),
+$$
+
+where $f_k$ represents the individual regression trees.
+
+The model is trained in an additive manner, where at each iteration, the objective function is approximated to efficiently compute the gain from adding a new tree. 
+The gain is determined by evaluating the improvement in accuracy from splitting a node into two new nodes. 
+It takes into account the gradients (first derivatives of the loss function) to evaluate the direction of optimization, and the Hessians (second derivatives of the loss function) to understand the curvature of the loss function.
```
Comment (Pattrigue) [inline report_thesis/src/sections/background.tex:306]: I think maybe this should be rephrased to avoid having the sentences in brackets

File: report_thesis/src/sections/background.tex [hunk 4]
```diff
@@ -286,6 +286,27 @@ \subsubsection{Natural Gradient Boosting (NGBoost)}
...
+The objective function to be minimized is composed of a convex loss function $L$ that measures the difference between the prediction $\hat{y}_i$ and the target $y_i$, and a regularization term $\Omega$ that penalizes the complexity of the model. 
+A convex loss function has the property of having a single global minimum, which simplifies the optimization process and ensures reliable convergence:
+
+$$
+Obj(\theta) = \sum_{i=1}^{n} L(y_i, \hat{y}_i) + \sum_{k=1}^{K} \Omega(f_k),
+$$
+
+where $f_k$ represents the individual regression trees.
+
+The model is trained in an additive manner where, at each iteration, the objective function is approximated to efficiently compute the gain from adding a new tree.
```
[the remark this answers is not in the document]
Comment (chhoumann) [reply report_thesis/src/sections/background.tex:304]: But is this right? I think it's not an exactly precise statement. I'm referring to:

> the objective function is approximated to efficiently compute the gain from adding a new tree. 

I'm not sure that's exactly true?

From what I can tell, the precise relationship is that the gain computation is a part of the process to optimize the objective function efficiently using the second-order approximation. Can you verify?

File: report_thesis/src/sections/background.tex [hunk 5]
```diff
@@ -286,6 +286,27 @@ \subsubsection{Natural Gradient Boosting (NGBoost)}
...
+A convex loss function has the property of having a single global minimum, which simplifies the optimization process and ensures reliable convergence:
+
+$$
+Obj(\theta) = \sum_{i=1}^{n} L(y_i, \hat{y}_i) + \sum_{k=1}^{K} \Omega(f_k),
+$$
+
+where $f_k$ represents the individual regression trees.
+
+The model is trained in an additive manner where, at each iteration, the objective function is approximated to efficiently compute the gain from adding a new tree. 
+The gain is determined by evaluating the improvement in accuracy from splitting a node into two new nodes.
```
Comment (chhoumann) [inline report_thesis/src/sections/background.tex:305]: Is accuracy right?

File: report_thesis/src/sections/background.tex (hunk 3 above)
Comment (chhoumann) [reply report_thesis/src/sections/background.tex:306]: Agreed. But could probably be simplified.

File: report_thesis/src/sections/background.tex [hunk 6]
```diff
@@ -286,6 +286,27 @@ \subsubsection{Natural Gradient Boosting (NGBoost)}
...
+$$
+Obj(\theta) = \sum_{i=1}^{n} L(y_i, \hat{y}_i) + \sum_{k=1}^{K} \Omega(f_k),
+$$
+
+where $f_k$ represents the individual regression trees.
+
+The model is trained in an additive manner where, at each iteration, the objective function is approximated to efficiently compute the gain from adding a new tree. 
+The gain is determined by evaluating the improvement in accuracy from splitting a node into two new nodes. 
+It takes into account the gradients (first derivatives of the loss function) to evaluate the direction of optimization, and the Hessians (second derivatives of the loss function) to understand the curvature of the loss function. 
+Additionally, the regularization parameter helps to control the complexity of the model, and a penalty term is included to account for the cost of adding new leaves to the tree.
```
Comment (chhoumann) [inline report_thesis/src/sections/background.tex:307]: Sure. This should probably be what we write about the regularization, instead of what's written earlier in the section.

**Review by chhoumann — CHANGES_REQUESTED** [review]
Need to check on some things before merging

File: report_thesis/src/sections/background.tex (hunk 6 above)
Comment (chhoumann) [reply report_thesis/src/sections/background.tex:307]: Referring to l307

File: report_thesis/src/sections/background.tex [hunk 7]
```diff
@@ -286,6 +286,27 @@ \subsubsection{Natural Gradient Boosting (NGBoost)}
...
 
 \subsubsection{XGBoost}
+We now describe XGBoost based on \citet{ChenGuestrin2016}.
+
+\gls{xgboost} is an advanced implementation of gradient boosting designed for speed and performance. 
+It builds upon traditional boosting techniques by introducing several algorithmic enhancements to improve execution speed and model performance.
+
+\gls{xgboost} optimizes the training process by using a more regularized model formalization to control overfitting. 
+The objective function to be minimized is composed of a convex loss function $L$ that measures the difference between the prediction $\hat{y}_i$ and the target $y_i$, and a regularization term $\Omega$ that penalizes the complexity of the model. 
+A convex loss function has the property of having a single global minimum, which simplifies the optimization process and ensures reliable convergence:
```
[the remark this answers is not in the document]
Comment (Ivikhostrup) [reply report_thesis/src/sections/background.tex:296]: Removed any mention of a convex loss function and just called it the loss function.

File: report_thesis/src/sections/background.tex [hunk 8]
```diff
@@ -286,6 +286,27 @@ \subsubsection{Natural Gradient Boosting (NGBoost)}
...
+
+$$
+Obj(\theta) = \sum_{i=1}^{n} L(y_i, \hat{y}_i) + \sum_{k=1}^{K} \Omega(f_k),
+$$
```
[the remark this answers is not in the document]
Comment (Ivikhostrup) [reply report_thesis/src/sections/background.tex:300]: Adjusted so that it matches 100%

File: report_thesis/src/sections/background.tex [hunk 9]
```diff
@@ -286,6 +286,27 @@ \subsubsection{Natural Gradient Boosting (NGBoost)}
...
+
+where $f_k$ represents the individual regression trees.
```
[the remark this answers is not in the document]
Comment (Ivikhostrup) [reply report_thesis/src/sections/background.tex:302]: Not worth the effort I think. This is a definition of xgboost and since its able to do multiple types of tasks, I think keeping it general serves well.

File: report_thesis/src/sections/background/ensemble_learning_models/xgboost.tex [hunk 10]
```diff
@@ -1 +1,27 @@
...
+	\end{itemize}
+\item This process is recursively applied to each partition, building a tree structure by continually splitting the data to minimize the loss function at each node.
+\end{enumerate}
+During tree construction, regularization is applied to control the complexity of the model. 
+The regularization term penalizes the complexity of the tree, helping to prevent overfitting.
+Once the tree is fully grown, a pruning step is performed. 
+Tree pruning works by starting from the leaves and moving backwards to the root, removing branches that do not contribute to reducing the overall loss function. 
+This step ensures that only the most useful splits are retained.
+Additionally, techniques such as shrinkage and column subsampling are employed. 
+When new weights are added shrinkage scales these weights by a factor $\eta$ after each step.
```
Comment (Pattrigue) [inline report_thesis/src/sections/background/ensemble_learning_models/xgboost.tex:25]:
```suggestion
When new weights are added, shrinkage scales these weights by a factor $\eta$ after each step.
```

File: report_thesis/src/sections/background/ensemble_learning_models/xgboost.tex [hunk 11]
```diff
@@ -1 +1,27 @@
...
+\item The algorithms first step involves sorting the feature values and proposing candidate split points based on the distribution of these values.
+\item It then maps these features into buckets based on the candidate splitting points.
+\item For each candidate split, the algorithm calculates a split score using a loss function that measures the difference between predicted and actual values.
+\item The split that minimizes the loss function is chosen, and the data is partitioned based on this optimal split point. 
+	\begin{itemize}
+		\item Parallelism is leveraged here, where the computations for different features and split points can be distributed across multiple processors. 
+		This is facilitated by storing each feature in a compressed, sorted column format, allowing efficient linear scans and parallel processing.
+	\end{itemize}
+\item This process is recursively applied to each partition, building a tree structure by continually splitting the data to minimize the loss function at each node.
+\end{enumerate}
```
Comment (Pattrigue) [inline report_thesis/src/sections/background/ensemble_learning_models/xgboost.tex:18]: I think this should be prose. It's written in prose (e.g. "the algorithm's first step"), but it's listed in a numerical list.

Generally, I have tried to avoid writing lists in the background section, instead opting to write it in prose and splitting steps with new lines.

File: report_thesis/src/sections/background/ensemble_learning_models/xgboost.tex [hunk 12]
```diff
@@ -1 +1,27 @@
...
+\subsubsection{XGBoost}
+We now describe XGBoost based on \citet{ChenGuestrin2016}.
+
+\gls{xgboost} is an advanced implementation of gradient boosting designed for speed and performance. 
+It builds upon traditional boosting techniques by introducing several algorithmic enhancements to improve execution speed and model performance.
+
+The core mechanism of \gls{xgboost} can be summarized in the following way:
+\begin{enumerate}
+\item The algorithms first step involves sorting the feature values and proposing candidate split points based on the distribution of these values.
+\item It then maps these features into buckets based on the candidate splitting points.
```
Comment (chhoumann) [inline report_thesis/src/sections/background/ensemble_learning_models/xgboost.tex:10]: Be careful here because it seems like you mix the approximate and exact greedy algorithms. They are different.

The exact greedy algorithm sorts by feature values and visits the data in sorted order to get gradient statistics for the split evaluation, whereas the approximate algorithm proposes candidate splits by percentiles of the feature distribution, then maps the continuous features to buckets split by the candidate points, aggregates the statistics, and then does the evaluation.

So it seems your step 1 and 2 are mixing up these 2 algorithms?

**Review by chhoumann — CHANGES_REQUESTED** [review]

File: report_thesis/src/sections/background/ensemble_learning_models/xgboost.tex (hunk 10 above)
[the remark this answers is not in the document]
Comment (Ivikhostrup) [reply report_thesis/src/sections/background/ensemble_learning_models/xgboost.tex:25]: Removed this. I also agree that its an ambigious point.

File: report_thesis/src/sections/background/ensemble_learning_models/xgboost.tex [hunk 13]
```diff
@@ -1 +1,23 @@
...
+
+The core mechanism of \gls{xgboost} can be summarized in the following way:
+The algorithm first sorts the feature values and visits the data in sorted order to evaluate all possible split by accumulating gradient statistics.
+Then, each possible split is evaluated using a regularized objective function. 
+This function incorporates both a loss function that measures the difference between predicted and actual values and a regularization term that penalizes the complexity of the model.
+This additional regularization term helps prevent overfitting by smoothening the final learned weights.
+For each candidate split, the algorithm calculates a split score based on the accumulated statistics. 
+The algorithm chooses the split with the best split score, which minimizes the regularized objective function.
+This chosen split it used to partition the data into two groups.
+This process is recursively applied to each partition, building a tree structure by continually splitting the data to minimize the regularized objective function at each node.
```
Comment (Ivikhostrup) [inline report_thesis/src/sections/background/ensemble_learning_models/xgboost.tex:15]: This is the only point not addressed by the paper, but since this is a general trait of tree based algorithms I suppose its fine to keep. 
Should I be more explicit about this?

**Comment by Ivikhostrup** [comment]
We decided that it was not worth describing xgboost since its just an extension of gbr with some extra tricks.
